Generate ChipError.java for Android platform

yhoyoon opened the pull request:
This change introduces a new Python script, `gen_chip_error.py`, which generates the `ChipError.java` file for the Android platform. The script processes header files containing error definitions and creates corresponding Java constants and documentation comments. It is based on the existing `error_table.py` script.

Additionally, this update includes the following changes:
- Update `.github/workflows/zap_templates.yaml` to include the generation of `ChipError.java` during the CI process.
- Exclude the generated `ChipError.java` file from Restyled processing by updating `.restyled.yaml`.
- Modify `AndroidChipPlatform-JNI.cpp` to add a JNI method for converting error codes to strings.

For example, the `ChipError` enum class can be used in the following way within the `onPairingComplete` callback function:

```kt
override fun onPairingComplete(errorCode: Long) {
  val chipError = ChipError.value(errorCode)
  logger.log(Level.INFO, "onPairingComplete with error code: $chipError")
  if (chipError != ChipError.CHIP_NO_ERROR) {
    setFailure("onPairingComplete failure")
  }
}
```

yhoyoon committed:
Generate ChipError.java for Android platform

This change introduces a new Python script, gen_chip_error.py, which
generates the ChipError.java file for the Android platform. The script
processes header files containing error definitions and creates
corresponding Java constants and documentation comments. It is based on
the existing error_table.py script.

Additionally, this update includes the following changes:
- Update .github/workflows/zap_templates.yaml to include the generation
  of ChipError.java during the CI process.
- Exclude the generated ChipError.java file from Restyled processing by
  updating .restyled.yaml.
- Modify AndroidChipPlatform-JNI.cpp to add a JNI method for converting
  error codes to strings.

For example, the ChipError class can be used in the following way within
the onPairingComplete callback function:

override fun onPairingComplete(errorCode: Long) {
  logger.log(Level.INFO, "onPairingComplete with error code: ${ChipError.toString(errorCode)}")
  if (errorCode != ChipError.CHIP_NO_ERROR) {
    setFailure("onPairingComplete failure")
  }
}

Signed-off-by: Youngho Yoon <34558998+yhoyoon@users.noreply.github.com>

restyled-commits committed:
Restyled by autopep8

yhoyoon committed:
Fix the rrors of ruff Python linter

Run ruff check
  ruff check
  shell: sh -e {0}
scripts/gen_chip_error.py:114:11: F541 [*] f-string without any placeholders
scripts/gen_chip_error.py:116:11: F541 [*] f-string without any placeholders
scripts/gen_chip_error.py:118:11: F541 [*] f-string without any placeholders
scripts/gen_chip_error.py:123:19: F541 [*] f-string without any placeholders
scripts/gen_chip_error.py:125:19: F541 [*] f-string without any placeholders
Found 5 errors.
[*] 5 fixable with the `--fix` option.
Error: Process completed with exit code 1.

Signed-off-by: Youngho Yoon <34558998+yhoyoon@users.noreply.github.com>

joonhaengHeo commented:
I think enum class would be more appropriate.

joonhaengHeo commented:
And, this class can be used in Java controller and kotlin controller. It is necessary to review whether the current location is appropriate.

@yunhanw-google @yufengwangca 
Please check this PR.

yhoyoon commented:
> I think enum class would be more appropriate.

Thank you, I will change the ChipError to enum class.

> And, this class can be used in Java controller and kotlin controller. It is necessary to review whether the current location is appropriate.
> 
> @yunhanw-google @yufengwangca
> Please check this PR.

When reviewing the appropriate location, could you also consider [virtual-device-app](https://github.com/project-chip/connectedhomeip/tree/master/examples/virtual-device-app/android)? It is a commissionee app based on Android platform.

yhoyoon committed:
Change ChipError to enum class

Example usage:
override fun onPairingComplete(errorCode: Long) {
  val chipError = ChipError.value(errorCode)
  logger.log(Level.INFO, "onPairingComplete with error code: $chipError")
  if (chipError != ChipError.CHIP_NO_ERROR) {
    setFailure("onPairingComplete failure")
  }
}

Signed-off-by: Youngho Yoon <34558998+yhoyoon@users.noreply.github.com>

yhoyoon committed:
Change the enum constant type from int to long

Signed-off-by: Youngho Yoon <34558998+yhoyoon@users.noreply.github.com>

yunhanw-google commented on src/platform/android/java/chip/platform/ChipError.java:
```diff
@@ -0,0 +1,304 @@
+     *
+     * These errors are defined in src/protocols/interaction_model/StatusCodeList.h
+     */
+    SUCCESS(1280L),
```
could you migrate src/controller/java/src/chip/devicecontroller/model/Status.java, update file in https://github.com/project-chip/connectedhomeip/blob/master/src/protocols/interaction_model/StatusCodeList.h#L27C28-L27C91

yhoyoon replied:
I understand. Since the `src/controller/java/src/chip/devicecontroller/model/Status.java` file already exists, I will modify the script to update the existing file instead of adding IM status codes to the `ChipError` class.

yhoyoon replied:
@yunhanw-google , I just pushed a change. If this was your intention, I will update the remaining files as well.

- src/controller/python/chip/interaction_model/__init__.py
- src/controller/java/src/matter/controller/model/Status.kt

yunhanw-google commented on src/platform/android/java/chip/platform/ChipError.java:
```diff
@@ -0,0 +1,304 @@
+     *
+     * These errors are defined in src/protocols/interaction_model/StatusCodeList.h
+     */
+    SUCCESS(1280L),
```
why it is 1280L instead of 0L?

yhoyoon replied:
The reason for this is that IM Status Codes are represented by a separate range in `CHIP_ERROR`. As mentioned in the review above, I will delete the IM StatusCodeList from the `ChipError.java` file and update the `src/controller/java/src/chip/devicecontroller/model/Status.java` file instead.

https://github.com/project-chip/connectedhomeip/blob/708558881bf6e5b0dac3c09489124907927b6cd2/src/app/MessageDef/StatusIB.cpp#L138-L151

bzbarsky-apple commented on src/platform/android/java/chip/platform/ChipError.java:
```diff
@@ -0,0 +1,304 @@
+     *
+     * These errors are defined in src/lib/core/CHIPError.h
+     */
+    CHIP_NO_ERROR(0L),
```
Note that these error codes are subject to change, so I am hoping this is not intended as public API...

yunhanw-google replied:
@bzbarsky-apple  in darwin, how do we expose/express the matter error to the public?

bzbarsky-apple replied:
See https://github.com/project-chip/connectedhomeip/blob/master/src/darwin/Framework/CHIP/MTRError.h

and there's a mapping from internal error values to public API ones.

yhoyoon replied:
@bzbarsky-apple Thank you for sharing an example on the Darwin platform. Could you please provide a specific example to illustrate the advantage of mapping MTRErrors and CHIP_ERRORs?

For your information, the Android ChipDeviceController already exposes a CHIP_ERROR.
https://github.com/project-chip/connectedhomeip/blob/708558881bf6e5b0dac3c09489124907927b6cd2/src/controller/java/src/chip/devicecontroller/ChipDeviceController.java#L1824-L1837

bzbarsky-apple replied:
All I'm saying is that CHIP_ERROR names (especially) and values (if the old value stops being used for anything in the SDK) are subject to change as needed.  If you want to expose them in public API, then you should probably have a migration plan for when those things happen.

tcarmelveilleux replied:
@yhoyoon The recommendation is to:

- Not allocate specific IDs that match between CHIP_ERROR and the equivalent Java enum, just keep the names
- Make a factory class implementation that generates the opaque enum (i.e. no match between the C++ side and Java side based on numbers and NEVER remove an enum value) from the CHIP_ERROR. The implementation can change and be regenerated over time.

yhoyoon committed:
Seperate IM status code generation

Signed-off-by: Youngho Yoon <34558998+yhoyoon@users.noreply.github.com>

yhoyoon committed:
Use CharToStringUTF instead of NewStringUTF

Signed-off-by: Youngho Yoon <34558998+yhoyoon@users.noreply.github.com>

restyled-commits committed:
Restyled by autopep8

andy31415 commented:
@yhoyoon what do we do with this PR? It is quite old now and I am unsure about the path taken here when considering backwards compatibility. Can we close it?
